src/test/crimson/test_monc.cc

```diff
@@ -0,0 +1,72 @@
+#include <seastar/core/app-template.hh>
+#include "common/ceph_argparse.h"
+#include "crimson/common/config_proxy.h"
+#include "crimson/mon/MonClient.h"
+#include "crimson/net/SocketMessenger.h"
+
+using Config = ceph::common::ConfigProxy;
+using MonClient = ceph::mon::Client;
+
+static seastar::future<> test_monc()
+{
+  return ceph::common::sharded_conf().start().then([] {
+    std::vector<const char*> args;
+    std::string cluster;
+    std::string conf_file_list;
+    auto init_params = ceph_argparse_early_args(args,
+                                                CEPH_ENTITY_TYPE_CLIENT,
+                                                &cluster,
+                                                &conf_file_list);
+    auto& conf = ceph::common::local_conf();
+    conf->name = init_params.name;
+    conf->cluster = cluster;
+    return conf.parse_config_files(conf_file_list);
+  }).then([] {
+    return seastar::do_with(ceph::net::SocketMessenger{entity_name_t::OSD(0)},
+                            [](ceph::net::Messenger& msgr) {
+      auto& conf = ceph::common::local_conf();
+      if (conf->ms_crc_data) {
+        msgr.set_crc_data();
+      }
+      if (conf->ms_crc_header) {
+        msgr.set_crc_header();
+      }
+      return seastar::do_with(MonClient{conf->name, msgr},
+                              [&msgr](auto& monc) {
+        return monc.build_initial_map().then([&monc] {
+          return monc.load_keyring();
+        }).then([&msgr, &monc] {
+          return msgr.start(&monc);
+        }).then([&monc] {
+          return monc.authenticate(std::chrono::seconds{10});
+        }).finally([&monc] {
+          return monc.stop();
```

Comment:
i don't see the test calling `msgr.shutdown()`, and i wonder about potential races between that and `monc.stop()`. the messenger still has a pointer to `mon::Client` (via `Dispatcher*`) and may continue making calls to that interface. maybe those ms_ calls in `mon::Client` should have an early return if its been closed?

Reply:
@cbodley will add `msgr.shutdown()`. regarding to the racing i thought your latest gate change should address it, but it cannot. can we make the `pending_dispatch` a member variable of `SocketConnection` ? so we can wait for it in`SocketConnection::close()` ?

Reply:
that would help with calls to `ms_dispatch()` on connections closed by `mon::Client::close()` - but the messenger could still be making other calls like `ms_accept()` or `ms_connect()`, and it wouldn't help there

i think `mon::Client` itself would need to gate its outstanding futures with `mon::Client::close()`. and in general, any Dispatcher would need to do the same. how about something like a `GatedDispatcher` to encapsulate this and present a new gated interface? ie
```c++
class GatedDispatcher : public Dispatcher {
  seastar::gate gate;
 protected: // present a new protected interface to be overridden
  virtual seastar::future<> gated_dispatch(ConnectionRef conn, MessageRef m) = 0;
  ...

  seastar::future<> gated_close() { return gate.close(); }

 public: // implement public Dispatcher interface in terms of the gated interfaces
  seastar::future<> ms_dispatch(ConnectionRef conn, MessageRef m) override final {
    return seastar::with_gate(gate, [=] { return gated_dispatch(conn, m); });
  }
  ...
};
```
so `mon::Client` would inherit from `GatedDispatcher`, implement the `gated_` interfaces, and call `gated_close()` on `mon::Client::close()`. it's kind of messy, but at least it keeps that mess out of both the Messenger and mon::Client?

Reply:
@cbodley thanks for the suggestions! i need to ponder on this and then get back to you.

Reply:
@cbodley i think the major concern here is that the the dispatcher captured by `SocketMessenger` for dispatching the `ms_handle_connect()` and `ms_handle_accept()` calls could be dangling reference when the continuations are scheduled by reactor.

but, in our use case, we attach a single dispatcher to one or more messengers, and this dispatcher instance will outlive the messenger. and we already have a `gate` guarding `SocketMessenger`. so, to avoid adding yet another `with_*` around the inner dispatch call, can we enforce a rule:
 the dispatcher attached to a messenger should be alive until `Messenger::shutdown()` returns, in other words, the messenger will be able to call `Dispatcher`'s methods until `Messenger::shutdown()` returns.

in the case of `mon::Client::stop()`, what it does is just draining all the floating continuations, for instance, the ones scheduled by the timer, and close the connection to monitor. the `mon::Client` instance should still be able to serve as a qualified `Dispatcher`, in the sense that it won't abort or segfault when being notified with calls like `ms_dispatch()` or `ms_handle_reset()`. but i think we can tolerate with exceptions in this case.

what do you think?

Reply:
> in the case of `mon::Client::stop()`, what it does is just draining all the floating continuations, for instance, the ones scheduled by the timer, and close the connection to monitor. the `mon::Client` instance should still be able to serve as a qualified `Dispatcher`, in the sense that it won't abort or segfault when being notified with calls like `ms_dispatch()` or `ms_handle_reset()`. but i think we can tolerate with exceptions in this case.

that all sounds reasonable, yeah - but isn't that basically what the seastar::gate gives us? a point after which new calls will throw an exception?

Reply:
yeah, `gated_*()`, `gated_close()` and `seastar::gate` ensure us that there is no floating continuations in the wild after that point in a unified way. but we don't get this for free. aside from an extra layer of `Dispatcher` in the code, we need to

1. increment the count in `gate`, throw if it's closed
1. move-copy the continuation and its arguments before applying it
1. attach a continuation to decrement the counter

what i want is to avoid adding another `seastar::with_gate()` in the call stack, if we plant this `gate` before dispatch calls, then each call of `gated_*` method will need to go through the `gate`. and the reason we do this is just for the graceful shutdown, which only happens once in a the whole life of crimson-osd process. i 

currently, `SocketMessenger` already guards all dispatcher calls with `pending_dispatch`, so we can be sure that all `ms_*` calls will be concluded after `SocketMessenger::shutdown()`. if we can also ensure that all `ms_*` calls are still safe *after* `mon::Client::stop()` as you suggested in the [first comment](https://github.com/ceph/ceph/pull/23849#discussion_r218470423) in this conversation thread

> may continue making calls to that interface. maybe those ms_ calls in `mon::Client` should have an early return if its been closed

my argument of not adding `GatedDispatcher` is that i assume the `ms_*` calls are well-behaved  even after `mon::Client::stop()`. and they are allowed to throw exceptions if they are not happy with a closed connection being passed in.

the early return for `ms_*` calls is not able to guarantee that we will be referencing a "stopped" dispatcher in the "later" part of `ms_*` continuation chain after it survives the early check in `ms_*` call for the "closed" status of dispatcher. we could check the "closed" status in every step of `ms_*` call, but that'd be more annoying.  i'd trade it for the grand and unified `GatedDispatcher`.

if you think the well-behaved `ms_*` calls after `stop()` is too strong a requirement for implementing `Dispatcher`s, i am happy to harden it with `seastar::gate`. and yeah, in that case, `GatedDispatcher` is a very good compromise.

Reply:
> my argument of not adding GatedDispatcher is that i assume the ms_* calls are well-behaved even after mon::Client::stop(). and they are allowed to throw exceptions if they are not happy with a closed connection being passed in.

IMO there's no need to worry about a closed connection being passed in, if the `dispatch_gate` of the msgr becomes a member of `SocketConnection`. (This is same idea with [the second comment](https://github.com/ceph/ceph/pull/23849/files#r218827807). I just implemented the per-conn `dispatch_gate` in #24142, it is because only the `Protocol` class is calling `Dispatcher` interfaces, and I think it is reasonable...)

> the early return for ms_* calls is not able to guarantee that we will be referencing a "stopped" dispatcher in the "later" part of ms_* continuation chain after it survives the early check in ms_* call for the "closed" status of dispatcher. we could check the "closed" status in every step of ms_* call, but that'd be more annoying. i'd trade it for the grand and unified GatedDispatcher.

Yes, it is also reasonable to me to stop dispatching if the `Dispatcher` is stopped.

> if you think the well-behaved ms_* calls after stop() is too strong a requirement for implementing Dispatchers, i am happy to harden it with seastar::gate. and yeah, in that case, GatedDispatcher is a very good compromise.

It is also possible to instead track the number of registered msgrs of this `Dispatcher`. The `Messenger` need to unregister its `Dispatchers` before shutdown, then the `Dispatcher` just need to wait for the number of msgrs drop to 0 before `stop()` returned. This way, there will be minor performance impact I think.